pkg/timelock/timelock.go

```diff
@@ -128,166 +128,256 @@ func NewTimelockWorker(nodeURL, timelockAddress, callProxyAddress, privateKey st
 // Listen is the main function of a Timelock Worker, it subscribes to events using the ethClient
 // and targeting the contract address set.
 func (tw *Worker) Listen(ctx context.Context) error {
-	if err := ctx.Err(); err != nil {
+	c, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Log timelock-worker configuration.
+	tw.startLog()
+
+	// Handle OS signals to properly stop/kill the process.
+	stopCh := make(chan string)
+	defer close(stopCh)
+	go handleOSSignal(stopCh)
+
+	// Run the scheduler to add/del operations in a thread-safe way.
+	go tw.runScheduler(c)
+
+	// Retrieve historical logs.
+	history, err := tw.retrieveHistoricalLogs(c)
+	if err != nil {
+		tw.logger.Error().Err(err).Msg("failed to subscribe and process logs.")
 		return err
 	}
 
-	stopCh := make(chan string)
+	// Create a subscription and retrieve new logs asynchronously.
 	logCh := make(chan types.Log)
-	sigCh := make(chan os.Signal, 1)
-	signal.Notify(sigCh, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM)
+	defer close(logCh)
 
-	tw.updateSchedulerDelay(time.Duration(tw.pollPeriod) * time.Second)
-	go tw.runScheduler(ctx)
-	tw.startLog()
+	if err := tw.subscribeNewLogs(c, logCh); err != nil {
+		tw.logger.Error().Err(err).Msg("failed to subscribe and process logs.")
+		return err
+	}
 
-	// Shutdown gracefully based on OS interrupts.
-	go func() {
-		for {
-			handleOSSignal(<-sigCh, stopCh)
-		}
-	}()
+	// Main goroutine; processes old and new logs and handles cancellation.
+	tw.processLogs(c, history, logCh, stopCh, cancel)
+
+	// Block until all goroutines are done.
+	wg.Wait()
 
-	// FilterQuery to be feed to the subscription and FilterLogs.
-	query := ethereum.FilterQuery{
+	tw.dumpOperationStore(time.Now)
+
+	return nil
+}
+
+func (tw *Worker) setupFilterQuery() ethereum.FilterQuery {
+	return ethereum.FilterQuery{
 		Addresses: tw.address,
 		FromBlock: tw.fromBlock,
 	}
+}
 
-	tw.logger.Info().Msgf("Starting subscription")
-	// Create the new subscription with the predefined query.
+// subscribeAndProcessLogs creates a subscription and processes logs based on a filter query.
+func (tw *Worker) subscribeNewLogs(ctx context.Context, logCh chan types.Log) error {
+	query := tw.setupFilterQuery()
+
+	// SubscribeFilterLogs creates an asynchronous subscription to the events.
+	// It receives all the new events.
 	sub, err := tw.ethClient.SubscribeFilterLogs(ctx, query, logCh)
 	if err != nil {
+		tw.logger.Error().Msgf("unexpected error while creating subscription: %s", err.Error())
 		return err
 	}
+	defer sub.Unsubscribe()
```

Comment:
should only unsubscribe when the go routine closes, otherwise will unsub when `subscribeNewLogs` returns